File: Java/arrays/src/VectorTest.java
import org.junit.After;
import org.junit.Before;
import org.junit.Test;


public class VectorTest {

    static Vector vector = new Vector<String>(10);

    @Before
    public void setUp(){
        System.out.println("---------------------------------------");
        for(int i = 0; i < 8; i++){
            vector.add("test" + i);
        }
        vector.show();
        System.out.println("Vector length: " + vector.length);
        System.out.println("Initial vector size: " + vector.size());
        System.out.println("---------------------------------------");
    }

    @After
    public void report(){
        System.out.println("---------------------------------------");
        System.out.println(vector.toString());
        System.out.println("Last vector size: " + vector.size());
        System.out.println("Last Vector length: " + vector.length);
        vector.clear();
        System.out.println("---------------------------------------");
    }

    @Test
    public void ShouldLengthIncreaseWhenFunctionAddIsCalled(){
        System.out.println("Testing Add function");
        try{
            vector.add("testAdd");
        }catch (IndexOutOfBoundsException e){
            System.out.println("Not enough capacity");
        }
        assert vector.length > 0;
    }

    @Test
    public void ShouldPopReduceVectorLengthWhenCalled(){
        System.out.println("Testing Pop function");
        vector.remove(6);
        System.out.println("Showing result");
        vector.show();

        assert vector.length < 10;
        assert vector.get(6).equals("test7");
    }

    @Test
    public void ShouldGetFindValueByIndex(){
        System.out.println("Testing Get function for not null value return");
        System.out.println("Adding value on the next index");
        vector.add("testGet");
        try{
            assert vector.get(8).equals("testGet");
        }catch (IllegalArgumentException e){
            System.out.println("Invalid index");
        }
    }

    @Test
    public void ShouldGetReturnNullValue(){
        System.out.println("Testing Get function for null value return");
        System.out.println("Adding value on the first index");
        vector.add("testGet2");
        try{
            assert vector.get(9) == null;
        }catch (IllegalArgumentException e){
            System.out.println("Invalid index");
        }
    }

    @Test
    public void ShouldSearchFindAndReturnANotNullValue(){
        System.out.println("Testing search function to find test5");
        assert vector.search("test5");
    }

    @Test
    public void ShouldSearchFailsWhenSearchingNonInsertedValues(){
        System.out.println("Testing Search function to find test15");
        assert !vector.search("test15");
    }

    @Test
    public void ShouldSearchAndGetElementByIndex(){
        System.out.println("Testing findIndex() method");
        System.out.println("Trying to find test6 index");
        assert vector.findIndex("test6") == 6;
    }

    @Test
    public void ShouldInsertInIndexFunctionInsertInAnyDeservedIndex(){
        try{
            vector.insertInIndex(3,"TestDoll");
            vector.show();
            assert vector.get(3) == "TestDoll";

        }catch(IllegalArgumentException e){
            System.out.println("Invalid index");
        }
    }

    @Test
    public void ShouldIncreaseCapacityWithouLosingItValues(){
        System.out.println("Testing increaseCapacity() method");
        System.out.println("Increasing by 4");
        vector.increaseCapacity(4);
        System.out.println("Showing it values");
        vector.show();

        assert vector.size() == 12;
    }
}

File: Kotlin/arrays/src/VectorTest.kt
import Vector
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.junit.After
import org.junit.Before


@RunWith(JUnit4::class)
class VectorTest {
    companion object{
        private val vector = Vector<String>(10)
    }

    @Before
    fun setUp(){
        println("-------------------------------------------------")
        for(i: Int in 0..8){
            vector.add("test${i}")
        }
        vector.show()
        println("Vector length: ${vector.length}")
        println("Vector initial capacity: ${vector.pickSize()}")
        println("-------------------------------------------------")
    }

    @After
    fun report(){
        println(vector.toString())
        println("Last Vector length: ${vector.length}")
        println("Last Vector capacity: ${vector.pickSize()}")
        vector.clear()
        println("-------------------------------------------------")
    }

    @Test
    fun `should length increase when function add is called`() {
        println("Testing Add function")
        vector.add("testAdd")
        assertEquals(10, vector.length)
    }

    @Test
    fun `should make all values blank`(){
        println("Testing clear() method")
        vector.clear()
        for(i: Int in 0..vector.length-1){
            assertEquals(Unit, vector.get(i))
        }
        assertEquals(0, vector.length)
    }

    @Test
    fun `should function add insert one value next the another`(){
        println("Testing add() method")
        vector.add("testAdd")
        assertEquals("testAdd", vector.get(9))
    }

    @Test
    fun `should pop remove the element from array`(){
        println("Testing pop() method to remove ${vector.get(7)}")
        try{
            println("Removing it...")
            vector.remove(6)
            vector.show()
            assertEquals("test7", vector.get(6))
            assertEquals(8, vector.length)
        }catch (e: Vector.ItemNotExistInVectorException){
            println("Item not found or won't exist")
            fail()
        }

    }

    @Test
    fun `should search for an element and return true`(){
        println("Testing search() method to find test6 value")
        assertEquals(true, vector.search("test6"))
    }

    @Test
    fun `should search for an element and return false`(){
        println("Testing search() method to not find value test12")
        assertEquals(false, vector.search("test12"))
    }

    @Test
    fun `should search and get element index`(){
        println("Testing findIndex() method")
        println("Trying find test2 index")
        assertEquals(2, vector.findIndex("test2"))
    }

    @Test
    fun `should search an element and return boolean`(){
        println("Testing contains() method")
        assertEquals(true, vector.contains("test4"))
        assertEquals(false, vector.contains("test9"))
    }

    @Test
    fun `should always returns the last element`(){
        println("Testing if last() method will always show the last element")
        assertEquals("test8", vector.last())
        vector.insertInIndex(9, "testDummy")
        assertEquals("testDummy", vector.last())
        vector.insertInIndex(9,"test9")
        assertEquals("testDummy", vector.last())
        assertEquals(false, vector.contains(Unit.toString()))
    }

    @Test
    fun `should insert element in any deserved index`(){
        println("Testing insertInIndex function")
        vector.insertInIndex(3,"TestDoll")
        vector.show()
        assertEquals("TestDoll", vector.get(3))
    }

    @Test
    fun `should remove an element and rearrange the next itens to the new positions`(){
        println("Testing remove() method")
        vector.remove("test3")
        vector.show()
        assertEquals("test4", vector.get(3))
    }

    @Test
    fun `should increaseCapacity() increase elemets capacity without losing it values`(){
        println("Testing increaseCapacity() method")
        println("Increasing by 4")
        vector.increaseCapacity(newCapacity = 4)
        println("Showing it values")
        vector.show()
        assertEquals(14, vector.pickSize())
    }



}